### PR TITLE
Tests: bucket-policy parser tests, and bug fixes

### PR DESCRIPTION
**Pull request by hackintoshrao**
- Tests for bucket policy parser. 
- Bug fixes
  - Queryparam being a constant to a variable 
  - Enchancement to Resources validation from String.HasPreifx to Split approach.
  - Condition validation stricter. 
  - Enhancement to Conditions validation to avoid ambiguity in rules. 
  - Enhancement to Principals Validation.

### CODE REVIEW

**Comment by hackintoshrao** [comment]
Adding some cases.

File: bucket-policy-handlers.go [hunk 1]
```diff
@@ -177,7 +177,7 @@ func (api objectStorageAPI) PutBucketPolicyHandler(w http.ResponseWriter, r *htt
 	}
 
 	// Parse check bucket policy.
-	if s3Error := checkBucketPolicy(bucket, bucketPolicy); s3Error != ErrNone {
+	if s3Error := checkBucketResource(bucket, bucketPolicy); s3Error != ErrNone {
```
Comment (harshavardhana) [inline bucket-policy-handlers.go]: Keep this as checkBucketPolicyResources, - checkBucketResources has no meaning.

File: bucket-policy-parser.go [hunk 2]
```diff
@@ -130,52 +143,51 @@ func isValidPrincipals(principals []string) (err error) {
...
-	if !ok {
-		err = errors.New("Unsupported principal style found: ‘" + strings.Join(principals, " ") + "’, please validate your policy document.")
-		return err
 	}
 	return nil
 }
 
+// isValidConditions - are valid conditions.
 func isValidConditions(conditions map[string]map[string]string) (err error) {
+	findString := func(a string, list []string) bool {
```
Comment (harshavardhana) [inline bucket-policy-parser.go:160]: needs a comment.

File: bucket-policy-parser.go [hunk 3]
```diff
@@ -130,52 +143,52 @@ func isValidPrincipals(principals []string) (err error) {
...
-		}
-		if len(conditions["StringNotEquals"]) > 0 {
-			_, s3PrefixOK := conditions["StringNotEquals"]["s3:prefix"]
-			_, s3MaxKeysOK := conditions["StringNotEquals"]["s3:max-keys"]
-			if !s3PrefixOK && !s3MaxKeysOK {
-				err = fmt.Errorf("Unsupported condition keys found: ‘%s’, please validate your policy document.",
-					conditions["StringNotEquals"])
-				return err
+			conditionArray, ok := conditionKeyVal[key]
+			if ok {
```
Comment (harshavardhana) [inline bucket-policy-parser.go]: combine "if ok && findString(conditions[conditionType][key], conditionArray)"

**Pull request merged**